juju/juju
# Add a new logsink worker to handle logs for each model

Previously we had controller logging concerns in various places: model manager worker, core logger package, api server, juju agent. And we logged some things (model worker logs) to model files and other things (agent logs) to the common logsink (as well as mongo and syslog).

This PR introduces a log sink worker which manages the logging concerns in one place. We still need to move the controller logsink.log writer across, but for now the worker ensures that all log lines emitted for a given model are written to 3 places:
- mongo
- syslog (if configured)
- `/var/log/juju/models/<modelfilename>.log`

NB previously the model log file only contained the model worker logs, not any agent log entries so it was pretty useless as a source of model log lines.

The log sink is wired up as a dependency for the apiserver worker and model workers. The latter workers now have a bunch of code deleted and less manual creation of dependencies. The syslog references can also be removed from core logger, which now just manages handing out generic loggers for models.

There's also a split in where the apiserver log sink parameters live. The mongo related ones are moved to the log sink worker. The api server rate limiting ones applicable to the `/logsink` endpoint are retained in apiserver.

A few TODOs have been left behind. One of them is to clean up logs from the source controller after a model migration. We have never done this.

debug-log is still connected to the records in mongo. A future PR will wire this up to tail the model log file on disk instead. 

## QA steps

bootstrap, add a model
juju debug-log off both models works
ssh into the controller and check the content of the `/var/log/juju/models/xxx.log` files
also check the logsink.log file content

bootstrap a second controller
migrate the model
debug-log on the migrated model works
ssh into the second controller and check the on disk model log file

## Links

**Jira card:** JUJU-5486
